### PR TITLE
[Forms] Getting child type better for collections. Format/Example extensions. 

### PR DESCRIPTION
**Pull request by aequasi**
Solves #1184

### CODE REVIEW

**Comment by aequasi** [comment]
lmk if you want any changes

**Comment by aequasi** [comment]
I'm not entirely sure how to handle collections properly...


If the collection is of a simple type, it shouldnt need the `ref`, and the example should be simple, but if the collection is complex (like multiple sub forms), idk...

File: ModelDescriber/FormModelDescriber.php [hunk 1]
```diff
@@ -38,13 +44,16 @@ public function __construct(FormFactoryInterface $formFactory = null)
     public function describe(Model $model, Schema $schema)
     {
         if (method_exists(AbstractType::class, 'setDefaultOptions')) {
-            throw new \LogicException('symfony/form < 3.0 is not supported, please upgrade to an higher version to use a form as a model.');
+            throw new \LogicException(
+                'symfony/form < 3.0 is not supported, please upgrade to an higher version to use a form as a model.'
```
Comment (GuilhemN) [inline ModelDescriber/FormModelDescriber.php]: Please remove unrelated changes

File: ModelDescriber/FormModelDescriber.php [hunk 2]
```diff
@@ -38,13 +44,16 @@ public function __construct(FormFactoryInterface $formFactory = null)
...
+            );
         }
         if (null === $this->formFactory) {
             throw new \LogicException('You need to enable forms in your application to use a form as a model.');
         }
 
         $schema->setType('object');
+        $properties = $schema->getProperties();
```
Comment (GuilhemN) [inline ModelDescriber/FormModelDescriber.php]: Can be removed ?

File: ModelDescriber/FormModelDescriber.php [hunk 3]
```diff
@@ -61,128 +70,101 @@ private function parseForm(Schema $schema, FormInterface $form)
...
+            if ($config->hasOption('format') && !empty($config->getOption('format'))) {
+                $property->setFormat($config->getOption('format'));
+            }
+            if ($config->hasOption('example') && !empty($config->getOption('example'))) {
+                $property->setExample($config->getOption('example'));
+            }
+
             if ($config->getRequired()) {
-                $required = $schema->getRequired() ?? [];
+                $required   = $schema->getRequired() ?? [];
```
Comment (GuilhemN) [inline ModelDescriber/FormModelDescriber.php]: Please revert this

**Comment by aequasi** [comment]
Like that?

File: ModelDescriber/FormModelDescriber.php [hunk 4]
```diff
@@ -61,99 +67,72 @@ private function parseForm(Schema $schema, FormInterface $form)
...
                 }
 
-                if ($type->getInnerType() && ($formClass = get_class($type->getInnerType())) && !$this->isBuiltinType($formClass)) {
-                    //if form type is not builtin in Form component.
-                    $model = new Model(new Type(Type::BUILTIN_TYPE_OBJECT, false, $formClass));
-                    $property->setRef($this->modelRegistry->register($model));
-
-                    break;
-                }
+                $property->{$method}($value);
```
Comment (GuilhemN) [inline ModelDescriber/FormModelDescriber.php]: You can use `$property->merge()` instead :)

File: ModelDescriber/FormModelDescriber.php [hunk 5]
```diff
@@ -61,99 +67,72 @@ private function parseForm(Schema $schema, FormInterface $form)
     {
         $properties = $schema->getProperties();
 
+        /** @var FormInterface $child */
         foreach ($form as $name => $child) {
-            $config = $child->getConfig();
+            $config   = $child->getConfig();
```
Comment (GuilhemN) [inline ModelDescriber/FormModelDescriber.php]: Unrelated

File: ModelDescriber/FormModelDescriber.php [hunk 6]
```diff
@@ -61,99 +67,72 @@ private function parseForm(Schema $schema, FormInterface $form)
...
             $property = $properties->get($name);
             for ($type = $config->getType(); null !== $type; $type = $type->getParent()) {
                 $blockPrefix = $type->getBlockPrefix();
+                $property->setType($this->getPropertyType(get_class($type->getInnerType()), $config));
 
-                if ('text' === $blockPrefix) {
-                    $property->setType('string');
+                switch (true) {
```
Comment (GuilhemN) [inline ModelDescriber/FormModelDescriber.php]: Please revert this, using a switch looks less clear to me

File: Form/Extension/DocumentationExtension.php [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+use Symfony\Component\Form\Extension\Core\Type\FormType;
+use Symfony\Component\Form\FormBuilderInterface;
+use Symfony\Component\Form\FormInterface;
+use Symfony\Component\Form\FormView;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+/**
+ * @author Aaron Scherer <aequasi@gmail.com>
+ *
+ * DocumentationExtension Class
```
Comment (GuilhemN) [inline Form/Extension/DocumentationExtension.php]: Can be removed

File: Form/Extension/DocumentationExtension.php [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+class DocumentationExtension extends AbstractTypeExtension
+{
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $builder->setAttribute('documentation', $options['documentation']);
+    }
+
+    public function buildView(FormView $view, FormInterface $form, array $options)
+    {
+        $view->vars['documentation'] = $options['documentation'];
```
Comment (GuilhemN) [inline Form/Extension/DocumentationExtension.php]: Is this truly needed ?

File: ModelDescriber/FormModelDescriber.php (hunk 6 above)
Comment (aequasi) [reply ModelDescriber/FormModelDescriber.php]: Alright. I changed it to a `switch` because i though it was considerably more clear. Easy to revert though

File: Form/Extension/DocumentationExtension.php (hunk 8 above)
Comment (aequasi) [reply Form/Extension/DocumentationExtension.php]: I was just copying how you had made https://github.com/nelmio/NelmioApiDocBundle/blob/2.x/Form/Extension/DescriptionFormTypeExtension.php#L37

File: ModelDescriber/FormModelDescriber.php (hunk 4 above)
Comment (aequasi) [reply ModelDescriber/FormModelDescriber.php]: Is it okay that no validation is done on it?

File: Form/Extension/DocumentationExtension.php (hunk 8 above)
Comment (GuilhemN) [reply Form/Extension/DocumentationExtension.php]: Ok, I would remove it in this case :) If we find it usefull afterwards we'll still be able to readd it.

**Comment by aequasi** [comment]
Still not sure about the collection type

**Comment by GuilhemN** [comment]
I extracted the form extension in https://github.com/nelmio/NelmioApiDocBundle/pull/1234. Please do one change per PR to make it easier for us to review :)

**Comment by aequasi** [comment]
👍

**Comment by GuilhemN** [comment]
Could you rework this PR to only keep the change about collections please ? (I merged https://github.com/nelmio/NelmioApiDocBundle/pull/1234)

**Comment by rbrigot** [comment]
I'm really interested by this PR. @aequasi Are you still on this ?

**Comment by goetas** [comment]
here https://github.com/nelmio/NelmioApiDocBundle/pull/1309 is an alternative solution to the same problem
